### DISCUSSION TITLE
PiKaraoke won't launch on Pi 3 running Bullseye - Problem with chrome/chromedriver/selenium?

### DISCUSSION THREAD

**Discussion by buckmasterj** [vicwomg/pikaraoke#298] [General]
After following the readme for a fresh install on a raspberry pi 3 and then running ./pikaraoke.sh I keep getting the following errors:

```
[2023-12-27 18:14:54] INFO: Fetching available songs in: /home/pi/pikaraoke-songs/
[27/Dec/2023:18:14:561 ENGINE Bus STARTING [2023-12-27 18:14:56 ] INFO: [27/Dec/2023:18:14:561 ENGINE Bus STARTING
[27/Dec/2023:18:14:571 ENGINE Serving on http://0.0.0.0:5555/
[2023-12-27 18:14:57] INFO: [27/Dec/2023:18:14:57) ENGINE Serving on http://0.0.0.0:5555/
[27/Dec/2023:18:14:571 ENGINE Bus STARTED
[2023-12-27 18:14:57) INFO: [27/Dec/2023:18:14:571 ENGINE Bus STARTED
Traceback (nost recent call last):
File "/home/pl/pikaraoke/app.py", line 863, in <module> driver webdriver.Chrome (service-service, options-options)
File "/home/pi/pikaraoke/.venu/lib/python3.9/site-packages/selenlun/webdriver/chrone/webdriver.py", line 45, in __init__
super().__init_
Pile "/home/pi/pikaraoke/.venu/lib/python3.9/site-packages/seleniun/uebdriver/chroniun/webdriver.py", line 61, in_init_ super().init (command_executor executor, options options)
File "/home/pi/pikaraoke/.venu/lib/python3.9/site-packages/seleniun/webdriver/remote/webdriver.py", line 209, in_init__
self.start session(capabilities)
File response = self.execute(Command NEW SESSION, caps) ["value"]
"/home/pl/pikaraoke/.venu/lib/python3.9/site-packages/seleniun/uebdriver/remote/webdriver.py", line 293, in start session
File " self.error handler.check_response (response)
/home/pi/pikaraoke/.venu/lib/python3.9/site-packages/seleniun/webdriver/remote/webdriver.py", line 348, in execute
File "/home/pi/pikaraoke/.venu/lib/python3.9/site-packages/seleniun/webdriver/remote/errorhandler.py", line 229, In check response
raise exception_class(message, screen, stacktrace)
selenium.common.exceptions.Sessionfot CreatedException: Message: session not created: Chrome failed to start: exited normally.
(session not created: DevToolsActivePort file doesn't exist)
(The process started from chrone location /usr/bin/chroniun-browser is no longer running, so ChromeDriver is assuming that Chrome has crashed.)
Stacktrace:
10 0x000000b29454 <unknown> #1 0x0000007fb1c8 <unknown>
#2 0x000000831630 <unknown>
#3 0x00000082e2b4 <unknown>
#4 0x0000008818dc <unknown>
#5 0x000000875f64 <unknown>
16 0x00000083a73c unknown>
#7 0x00000083bdd0 <unknown>
#8 0x000000af32d4 <unknown>
#9 0x000000af7794 <unknown>
#10 0x000000af71c0 <unknown>
#11 0x000000af7bd4 <unknown>
#12 0x000000ae03c0 <unknown>
#13 0x000000af7fe8 <unknown>
#14 0x000000ac951c <unknown>
#15 0x000000b17ff8 <unknown>
#16 0x000000b18144 <unknown>
#17 0x000000b2861c <unknown>
#18 0x000076f50310 start_thread
#19 0x000000000000 <unknown>
```

I've been pulling my hair out trying to figure out what I did wrong?

I'm running Bullseye 32bit OS on a Pi3 model A+

**Comment by buckmasterj** [comment]
Got pretty much the same error after doing a fresh install on a Pi4 and running ./pikaraoke.sh

`Creating download path: /home/pi/pikaraoke-songs/
[2023-12-27 18:59:221 INFO: Fetching available songs in: /home/pi/pikaraoke-songs/
[27/Dec/2023:18:59:231 ENGINE Bus STARTING [2023-12-27 18:59:23) INFO: [27/Dec/2023:18:59:231 ENGINE Bus STARTING
[27/Dec/2023:18:59:231 ENGINE Serving on http://0.0.0.0:5555
[2023-12-27 18:59:23) INFO: [27/Dec/2023:18:59:231 ENGINE Serving on http://0.0.0.0:5555
[27/Dec/2023:18:59:23) ENGINE Bus STARTED [2023-12-27 18:59:231 INFO: [27/Dec/2023:18:59:231 ENGINE Bus STARTED
Traceback (nost recent call last):
File "/home/pl/pikaraoke/app.py", line 863, in <module>
driver webdriver.Chrome (service-service, options options)
File "/home/pl/pikaraoke/.venu/lib/python3.11/site-packages/selenium/webdriver/chrome/uebdriver.py", line 45, in_init_
super().init__( File "/home/pl/pikaraoke/.venu/lib/python3.11/site-packages/selenium/webdriver/chromiun/webdriver.py", line 61, in __init_
super()._init_(command_executor executor, options-options) /home/pi/pikaranke/.venu/lib/python3.11/site-packages/seleniun/webdriver/remote/vebdriver.py", line 209, in_init__
File "
self.start session(capabilities)
File "/home/pl/pikaraoke/.venv/lib/python3.11/site-packages/seleniun/webdriver/remote/webdriver.py", line 293, in start session
response = self.execute(Command. NEM SESSION, caps) ["value"]
File "/home/pi/pikaraoke/.venv/lib/python3.11/site-packages/seleniun/webdriver/renote/webdriver.py ", line 348, in execute
self.error handler.check_response (response)
File "/home/pl/pikaraoke/.venv/lib/python3.11/site-packages/seleniun/webdriver/remote/errorhandler.py", line 229, in check response
raise exception_class(message, screen, stacktrace)
seleniun.common.exceptions.SessionHot CreatedException: Message: session not created: Chrome failed to start: exited normally.
(session not created: DevToolsActivePort file doesn't exist)
(The process started fron chrome location /usr/bin/chronium-browser is no longer running, so Chromedriver is assuming that Chrome has crashed.)
Stacktrace:
10 0x0055901175f4 <unknown>
#1 0x00558fe07264 <unknown>
#2 0x00558fe39304 <unknown>
#3 0x00558fe357cc <unknown>
#4 0x00558fe7bade <unknown>
#5 0x00558fe40aZc <unknown>
#6 0x00558fe41c44 <unknown>
#7 0x0055900e160c <unknown>
#8 0x0055900c49c0 <unknown>
#9 0x0055900d1aec <unknown>
#10 0x0055900c537c <unknown>
#11 0x0055900b9eae <unknown> #12 0x005590195310 <unknown>
#13 0x0055901055bc <unknown>
#14 0x0055901167a8 <unknown>
#15 0x007f9af9ee18 <unknown> #16 0x007f9af77e9c <unknown>`

**Comment by vicwomg** [comment]
Did you run ./setup.sh first? That is a required step. What was the output
of that? Also make sure you’re NOT running these as sudo

There’s something unusual about this log output. “uebdriver.py” should be
“webdriver.py”  “chronium” should be “chromium”.  Did you use some kind of
image to text OCR to grab this text?

**Reply by buckmasterj** [reply]
Hi, yes I did run setup.sh first and it completed without any errors.

And yes, the OCR didn't do that great a job with the text

**Reply by buckmasterj** [reply]
I've tried on fresh RasPi OS image using the Raspberry Pi imager. 

After raspi-config to set up wifi, here are the steps I've been following:

Sudo apt-get update
git clone https://github.com/vicwomg/pikaraoke.git
cd pikaraoke
./setup.sh
./pikaraoke.sh

Also tried on a raspberry pi 5 with no luck. I must be missing some step?

**Comment by vicwomg** [comment]
Yup you’re doing everything right. Hm I wonder if something has changed
with the latest version of chromium that might have broken something.

Try running ./pikaraoke.sh —headless

That will skip the selenium auto-launching launching part. Then manually
connect to the url it specifies. Does that work?

**Comment by vicwomg** [comment]
Also, are you stuck on bullseye? Can you try raspberry pi os bookworm? All
of the release testing occurred there, so I’m unsure if bullseye works.

Also, Are you running desktop? Note that desktop is required since
pikaraoke needs a web browser.

**Reply by buckmasterj** [reply]
Will try bookworm, thanks

**Reply by buckmasterj** [reply]
It does load on bookworm! but it absolutely crawls on a pi3 model a. I am also getting the following message before it fully loads:
"Due to limitations with browser permissions, you must interact with the page once before it allows autoplay of videos. Pikaraoke will not play otherwise. click the button below to confirm"
The song did download but it took 10+ minutes (I'm on a fiber connnection) and is now stopping and starting the music but no video/lyrics showing.

Writing a new image to try on a Pi4

**Reply by vicwomg** [reply]
The "you must interact with the page" message is expected and it's true, you DO have to click on a browser screen before it can play video, this is a restriction on all browsers meant to prevent unwanted media such as ads. However, the workaround is if you launch pikaraoke headed mode Selenium fires up the browser and clicks this button for you. If you run headless, the remote browser will have to do this manually. 

I personally have run on a Pi 3 Model B just fine. It struggles a bit with user-supplied higher res videos, but nothing that comes from yt-dlp will be higher than that threshold unless you unless you force it to with the CLI option.

IIRC had to make a few tweaks to get it working smoothly on the pi3 (which is explained a tiny bit in the README, maybe should be more visible)
- Use 32-bit version of bookworm, 64-bit seemed worse 
- Overclock to `arm_freq=1300`

Pi 4 should work more or less without issue on 32 or 64 bit with no modification

**Reply by buckmasterj** [reply]
Ok thanks! Never tried the OC, maybe that will do the trick on the pi3. Good news is it fired up and runs great on the pi 4! Just set it up. Bookworm was the sticking point I guess? Thanks again

**Reply by vicwomg** [reply]
Yes, I guess there's an incompatibility with Bullseye, not sure why and I'm afraid I don't have the resources to look into it for a while.

For now, I'll add that Bookworm is reccommened to the README.md but would be good to figure out the Bullseye+Selenium incompatibility eventually for someone who has time.

**Comment by Burlanda** [comment]
Hi! Great project! Many thanks for it. 
I have been using ver 1.1.2 for an year on a RPI4 with retropie 4.8 distribution. Amazing! A few days ago I noticed some hickups in downloading. Youtube-dl was sometimes trowing an error about compatability with python 3.7 (is it possible?). Anyway, I also tried ver. 1.2. and I got some problems. When launching ./pikaraoke.sh or sudo pikaraoke.sh or python3 app.py it gave an error: "if (path := os.getenv("SE_MANAGER_PATH")) is not None: ^ SyntaxError: invalid syntax" 
Also I noticed that python3-venv needs to be preinstalled using sudo apt intall python3-venv - similar to pip3...
Another minor thing in ver 1.2, the file constants.py at line 1 reports: VERSION = "1.1.2" ... should it not be 1.2?
So, I got back to version 1.1.2, but eventually I will have to migrate to a new version when retropie will be also update.

**Reply by vicwomg** [reply]
Yes, this is a python compatibility issue. Python 3.7 does not support the "walrus operator" `:=`. Updating to the latest version of Python 3 should fix it.

**Comment by lsibarra** [comment]
Hi,
I am also having issues with this. I already switched it to the Bookworm OS in Raspberry Pi 3. Just to be clear I installed Bookworm with Desktop and Recommended software. However after running ./pikaraoke.sh --prefer-ip I am still getting the selenium chromedriver issue. See below.

`pi@raspberrypi:~/pikaraoke $ ./pikaraoke.sh --prefer-ip
[2024-01-05 18:28:32] INFO: Fetching available songs in: /home/pi/pikaraoke-songs/
[05/Jan/2024:18:28:34] ENGINE Bus STARTING
[2024-01-05 18:28:34] INFO: [05/Jan/2024:18:28:34] ENGINE Bus STARTING
[05/Jan/2024:18:28:34] ENGINE Serving on http://0.0.0.0:5555
[2024-01-05 18:28:34] INFO: [05/Jan/2024:18:28:34] ENGINE Serving on http://0.0.0.0:5555
[05/Jan/2024:18:28:34] ENGINE Bus STARTED
[2024-01-05 18:28:34] INFO: [05/Jan/2024:18:28:34] ENGINE Bus STARTED
Traceback (most recent call last):
  File "/home/pi/pikaraoke/app.py", line 863, in <module>
    driver = webdriver.Chrome(service=service, options=options)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/pi/pikaraoke/.venv/lib/python3.11/site-packages/selenium/webdriver/chrome/webdriver.py", line 45, in __init__
    super().__init__(
  File "/home/pi/pikaraoke/.venv/lib/python3.11/site-packages/selenium/webdriver/chromium/webdriver.py", line 61, in __init__
    super().__init__(command_executor=executor, options=options)
  File "/home/pi/pikaraoke/.venv/lib/python3.11/site-packages/selenium/webdriver/remote/webdriver.py", line 209, in __init__
    self.start_session(capabilities)
  File "/home/pi/pikaraoke/.venv/lib/python3.11/site-packages/selenium/webdriver/remote/webdriver.py", line 293, in start_session
    response = self.execute(Command.NEW_SESSION, caps)["value"]
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/pi/pikaraoke/.venv/lib/python3.11/site-packages/selenium/webdriver/remote/webdriver.py", line 348, in execute
    self.error_handler.check_response(response)
  File "/home/pi/pikaraoke/.venv/lib/python3.11/site-packages/selenium/webdriver/remote/errorhandler.py", line 229, in check_response
    raise exception_class(message, screen, stacktrace)
selenium.common.exceptions.SessionNotCreatedException: Message: session not created: Chrome failed to start: exited normally.
  (session not created: DevToolsActivePort file doesn't exist)
  (The process started from chrome location /usr/bin/chromium-browser is no longer running, so ChromeDriver is assuming that Chrome has crashed.)
Stacktrace:
#0 0x000000b59454 <unknown>
#1 0x00000082b1c8 <unknown>
#2 0x000000861b30 <unknown>
#3 0x00000085e2b4 <unknown>
#4 0x0000008b18dc <unknown>
#5 0x0000008a5f64 <unknown>
#6 0x00000086a73c <unknown>
#7 0x00000086bdd0 <unknown>
#8 0x000000b232d4 <unknown>
#9 0x000000b27794 <unknown>
#10 0x000000b271c0 <unknown>
#11 0x000000b27bd4 <unknown>
#12 0x000000b103c0 <unknown>
#13 0x000000b27fe8 <unknown>
#14 0x000000af951c <unknown>
#15 0x000000b47ff8 <unknown>
#16 0x000000b48144 <unknown>
#17 0x000000b5861c <unknown>
#18 0x000076b018e0 <unknown>
#19 0x000000000000 <unknown>
`

Here is the Chromium browser version and Chrome Driver

`
pi@raspberrypi:~/pikaraoke $ chromedriver --version
ChromeDriver 120.0.6099.102 (3660535ab30e616081ffe78f013a1852eeacfbc0-refs/branch-heads/6099@{#1412})
pi@raspberrypi:~/pikaraoke $ chromium-browser --version
Chromium 120.0.6099.102 Built on Raspbian , running on Raspbian 12
`

Raspbian OS version

`
pi@raspberrypi:~/pikaraoke $ cat /etc/os-release
PRETTY_NAME="Raspbian GNU/Linux 12 (bookworm)"
NAME="Raspbian GNU/Linux"
VERSION_ID="12"
VERSION="12 (bookworm)"
VERSION_CODENAME=bookworm
ID=raspbian
ID_LIKE=debian
HOME_URL="http://www.raspbian.org/"
SUPPORT_URL="http://www.raspbian.org/RaspbianForums"
BUG_REPORT_URL="http://www.raspbian.org/RaspbianBugs"
`

**Reply by vicwomg** [reply]
How are you launching this? Is it over ssh? If so, the terminal probably doesn't have access to the desktop window server. You may want to try launching directly on the pi

**Reply by lsibarra** [reply]
Hi @vicwomg,
Thanks for your guidance. I ran the command in the Raspberry Pi Bookworm OS and it worked. Thanks again for all your help on this.

**Comment by lsibarra** [comment]
Hi @vicwomg 
I had to reinstall my the pikaraoke on my rapsberry pi 3 because I needed a larger microsd card. I followed all the same steps. but now if I launch `./pikaraoke.sh --prefer-ip` it is no longer working. Looking at the Usage documented on github, it does not look like `--prefer-ip` is available? Is this no longer available? How do you recommend me launching ./pikaraoke.sh?

Thanks in advance.

-Larry

**Comment by vicwomg** [comment]
Prefer ip is now default. No need to specify it

**Reply by lsibarra** [reply]
HI @vicwomg 
Thanks for your reply. Was there a recent update after 01/05/2024 to the 1.2 version? I tried to install this and somehow the chromedriver is not getting installed in the raspberry pi after running `./setup.sh` Therefore the selenium is failing.

**Comment by vicwomg** [comment]
Include Output of the setup command please

**Comment by lsibarra** [comment]
Hi @vicwomg ,

Here is the log output after running `./setup.sh`

`pi@pikaraoke:~/pikaraoke $ ./setup.sh
Are you sure you want to setup PiKaraoke? (y/n): y
Client is not Debian-based. Skipping binary installation. Please install ffmpeg and chrome manually.

*** CREATING PYTHON VIRTUAL ENVIRONMENT ***

*** INSTALLING PYTHON DEPENDENCIES ***
Looking in indexes: https://pypi.org/simple, https://www.piwheels.org/simple
Collecting CherryPy
  Using cached https://www.piwheels.org/simple/cherrypy/CherryPy-18.9.0-py3-none-any.whl (348 kB)
Collecting Flask==2.2.5
  Using cached https://www.piwheels.org/simple/flask/Flask-2.2.5-py3-none-any.whl (101 kB)
Collecting qrcode
  Using cached https://www.piwheels.org/simple/qrcode/qrcode-7.4.2-py3-none-any.whl (46 kB)
Collecting psutil
  Using cached https://www.piwheels.org/simple/psutil/psutil-5.9.7-cp311-abi3-linux_armv7l.whl (280 kB)
Collecting unidecode
  Using cached https://www.piwheels.org/simple/unidecode/Unidecode-1.3.8-py3-none-any.whl (235 kB)
Collecting requests
  Using cached https://www.piwheels.org/simple/requests/requests-2.31.0-py3-none-any.whl (62 kB)
Collecting flask-paginate==2021.10.29
  Using cached https://www.piwheels.org/simple/flask-paginate/flask_paginate-2021.10.29-py2.py3-none-any.whl (7.3 kB)
Collecting jinja2==3.0.3
  Using cached https://www.piwheels.org/simple/jinja2/Jinja2-3.0.3-py3-none-any.whl (133 kB)
Collecting Babel==2.9.1
  Using cached https://www.piwheels.org/simple/babel/Babel-2.9.1-py2.py3-none-any.whl (8.8 MB)
Collecting Flask-Babel==2.0.0
  Using cached https://www.piwheels.org/simple/flask-babel/Flask_Babel-2.0.0-py3-none-any.whl (9.3 kB)
Collecting selenium
  Using cached https://www.piwheels.org/simple/selenium/selenium-4.16.0-py3-none-any.whl (1.1 MB)
Collecting ffmpeg-python
  Using cached https://www.piwheels.org/simple/ffmpeg-python/ffmpeg_python-0.2.0-py3-none-any.whl (25 kB)
Collecting yt-dlp
  Using cached https://www.piwheels.org/simple/yt-dlp/yt_dlp-2023.12.30-py2.py3-none-any.whl (3.0 MB)
Collecting Werkzeug>=2.2.2
  Using cached https://www.piwheels.org/simple/werkzeug/werkzeug-3.0.1-py3-none-any.whl (226 kB)
Collecting itsdangerous>=2.0
  Using cached https://www.piwheels.org/simple/itsdangerous/itsdangerous-2.1.2-py3-none-any.whl (15 kB)
Collecting click>=8.0
  Using cached https://www.piwheels.org/simple/click/click-8.1.7-py3-none-any.whl (97 kB)
Collecting MarkupSafe>=2.0
  Using cached https://www.piwheels.org/simple/markupsafe/MarkupSafe-2.1.3-cp311-cp311-linux_armv7l.whl (26 kB)
Collecting pytz>=2015.7
  Using cached https://www.piwheels.org/simple/pytz/pytz-2023.3.post1-py3-none-any.whl (502 kB)
Collecting cheroot>=8.2.1
  Using cached https://www.piwheels.org/simple/cheroot/cheroot-10.0.0-py3-none-any.whl (101 kB)
Collecting portend>=2.1.1
  Using cached https://www.piwheels.org/simple/portend/portend-3.2.0-py3-none-any.whl (5.3 kB)
Collecting more-itertools
  Using cached https://www.piwheels.org/simple/more-itertools/more_itertools-10.2.0-py3-none-any.whl (57 kB)
Collecting zc.lockfile
  Using cached https://www.piwheels.org/simple/zc-lockfile/zc.lockfile-3.0.post1-py3-none-any.whl (9.8 kB)
Collecting jaraco.collections
  Using cached https://www.piwheels.org/simple/jaraco-collections/jaraco.collections-5.0.0-py3-none-any.whl (10 kB)
Collecting typing-extensions
  Using cached https://www.piwheels.org/simple/typing-extensions/typing_extensions-4.9.0-py3-none-any.whl (32 kB)
Collecting pypng
  Using cached https://www.piwheels.org/simple/pypng/pypng-0.20220715.0-py3-none-any.whl (58 kB)
Collecting charset-normalizer<4,>=2
  Using cached https://www.piwheels.org/simple/charset-normalizer/charset_normalizer-3.3.2-py3-none-any.whl (48 kB)
Collecting idna<4,>=2.5
  Using cached https://www.piwheels.org/simple/idna/idna-3.6-py3-none-any.whl (61 kB)
Collecting urllib3<3,>=1.21.1
  Using cached https://www.piwheels.org/simple/urllib3/urllib3-2.1.0-py3-none-any.whl (104 kB)
Collecting certifi>=2017.4.17
  Using cached https://www.piwheels.org/simple/certifi/certifi-2023.11.17-py3-none-any.whl (162 kB)
Collecting trio~=0.17
  Using cached https://www.piwheels.org/simple/trio/trio-0.24.0-py3-none-any.whl (460 kB)
Collecting trio-websocket~=0.9
  Using cached https://www.piwheels.org/simple/trio-websocket/trio_websocket-0.11.1-py3-none-any.whl (17 kB)
Collecting future
  Using cached https://www.piwheels.org/simple/future/future-0.18.3-py3-none-any.whl (492 kB)
Collecting mutagen
  Using cached https://www.piwheels.org/simple/mutagen/mutagen-1.47.0-py3-none-any.whl (194 kB)
Collecting pycryptodomex
  Using cached https://www.piwheels.org/simple/pycryptodomex/pycryptodomex-3.20.0-cp35-abi3-linux_armv7l.whl (1.8 MB)
Collecting websockets>=12.0
  Using cached https://www.piwheels.org/simple/websockets/websockets-12.0-cp311-cp311-linux_armv7l.whl (129 kB)
Collecting brotli
  Using cached https://www.piwheels.org/simple/brotli/Brotli-1.1.0-cp311-cp311-linux_armv7l.whl (2.8 MB)
Collecting jaraco.functools
  Using cached https://www.piwheels.org/simple/jaraco-functools/jaraco.functools-4.0.0-py3-none-any.whl (9.8 kB)
Collecting tempora>=1.8
  Using cached https://www.piwheels.org/simple/tempora/tempora-5.5.0-py3-none-any.whl (13 kB)
Collecting attrs>=20.1.0
  Using cached https://www.piwheels.org/simple/attrs/attrs-23.2.0-py3-none-any.whl (60 kB)
Collecting sortedcontainers
  Using cached https://www.piwheels.org/simple/sortedcontainers/sortedcontainers-2.4.0-py2.py3-none-any.whl (29 kB)
Collecting outcome
  Using cached https://www.piwheels.org/simple/outcome/outcome-1.3.0.post0-py2.py3-none-any.whl (10 kB)
Collecting sniffio>=1.3.0
  Using cached https://www.piwheels.org/simple/sniffio/sniffio-1.3.0-py3-none-any.whl (10 kB)
Collecting wsproto>=0.14
  Using cached https://www.piwheels.org/simple/wsproto/wsproto-1.2.0-py3-none-any.whl (24 kB)
Collecting pysocks!=1.5.7,<2.0,>=1.5.6
  Using cached https://www.piwheels.org/simple/pysocks/PySocks-1.7.1-py3-none-any.whl (16 kB)
Collecting jaraco.text
  Using cached https://www.piwheels.org/simple/jaraco-text/jaraco.text-3.12.0-py3-none-any.whl (11 kB)
Requirement already satisfied: setuptools in ./.venv/lib/python3.11/site-packages (from zc.lockfile->CherryPy->-r requirements.txt (line 1)) (66.1.1)
Collecting h11<1,>=0.9.0
  Using cached https://www.piwheels.org/simple/h11/h11-0.14.0-py3-none-any.whl (58 kB)
Collecting jaraco.context>=4.1
  Using cached https://www.piwheels.org/simple/jaraco-context/jaraco.context-4.3.0-py3-none-any.whl (5.3 kB)
Collecting autocommand
  Using cached https://www.piwheels.org/simple/autocommand/autocommand-2.2.2-py3-none-any.whl (19 kB)
Collecting inflect
  Using cached https://www.piwheels.org/simple/inflect/inflect-7.0.0-py3-none-any.whl (34 kB)
WARNING: Retrying (Retry(total=4, connect=None, read=None, redirect=None, status=None)) after connection broken by 'ProtocolError('Connection aborted.', RemoteDisconnected('Remote end closed connection without response'))': /simple/pydantic/
Collecting pydantic>=1.9.1
  Using cached https://www.piwheels.org/simple/pydantic/pydantic-2.5.3-py3-none-any.whl (381 kB)
Collecting annotated-types>=0.4.0
  Using cached https://www.piwheels.org/simple/annotated-types/annotated_types-0.6.0-py3-none-any.whl (12 kB)
Collecting pydantic-core==2.14.6
  Using cached pydantic_core-2.14.6-cp311-cp311-manylinux_2_17_armv7l.manylinux2014_armv7l.whl (1.9 MB)
Installing collected packages: sortedcontainers, pytz, pypng, brotli, zc.lockfile, websockets, urllib3, unidecode, typing-extensions, sniffio, pysocks, pycryptodomex, psutil, mutagen, more-itertools, MarkupSafe, jaraco.context, itsdangerous, idna, h11, future, click, charset-normalizer, certifi, Babel, autocommand, attrs, annotated-types, wsproto, Werkzeug, requests, qrcode, pydantic-core, outcome, jinja2, jaraco.functools, ffmpeg-python, yt-dlp, trio, tempora, pydantic, Flask, cheroot, trio-websocket, portend, inflect, flask-paginate, Flask-Babel, selenium, jaraco.text, jaraco.collections, CherryPy
Successfully installed Babel-2.9.1 CherryPy-18.9.0 Flask-2.2.5 Flask-Babel-2.0.0 MarkupSafe-2.1.3 Werkzeug-3.0.1 annotated-types-0.6.0 attrs-23.2.0 autocommand-2.2.2 brotli-1.1.0 certifi-2023.11.17 charset-normalizer-3.3.2 cheroot-10.0.0 click-8.1.7 ffmpeg-python-0.2.0 flask-paginate-2021.10.29 future-0.18.3 h11-0.14.0 idna-3.6 inflect-7.0.0 itsdangerous-2.1.2 jaraco.collections-5.0.0 jaraco.context-4.3.0 jaraco.functools-4.0.0 jaraco.text-3.12.0 jinja2-3.0.3 more-itertools-10.2.0 mutagen-1.47.0 outcome-1.3.0.post0 portend-3.2.0 psutil-5.9.7 pycryptodomex-3.20.0 pydantic-2.5.3 pydantic-core-2.14.6 pypng-0.20220715.0 pysocks-1.7.1 pytz-2023.3.post1 qrcode-7.4.2 requests-2.31.0 selenium-4.16.0 sniffio-1.3.0 sortedcontainers-2.4.0 tempora-5.5.0 trio-0.24.0 trio-websocket-0.11.1 typing-extensions-4.9.0 unidecode-1.3.8 urllib3-2.1.0 websockets-12.0 wsproto-1.2.0 yt-dlp-2023.12.30 zc.lockfile-3.0.post1

*** DONE ***
Run PiKaraoke with: ./pikaraoke.sh <args>

pi@pikaraoke:~/pikaraoke $`

`pi@pikaraoke:~/pikaraoke $ chromedriver --version
-bash: chromedriver: command not found`

`pi@pikaraoke:~/pikaraoke $ chromium-browser --version
Chromium 119.0.6045.171 Built on Raspbian , running on Raspbian 12`

After the setup, running ./pikaraoke.sh I get this output

`pi@pikaraoke:~/pikaraoke $ ./pikaraoke.sh
[2024-01-17 22:12:29] INFO: Fetching available songs in: /home/pi/pikaraoke-songs/
[17/Jan/2024:22:12:30] ENGINE Bus STARTING
[2024-01-17 22:12:30] INFO: [17/Jan/2024:22:12:30] ENGINE Bus STARTING
[17/Jan/2024:22:12:31] ENGINE Serving on http://0.0.0.0:5555
[2024-01-17 22:12:31] INFO: [17/Jan/2024:22:12:31] ENGINE Serving on http://0.0.0.0:5555
[17/Jan/2024:22:12:31] ENGINE Bus STARTED
[2024-01-17 22:12:31] INFO: [17/Jan/2024:22:12:31] ENGINE Bus STARTED
Traceback (most recent call last):
  File "/home/pi/pikaraoke/app.py", line 847, in <module>
    driver = webdriver.Chrome(service=service, options=options)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/pi/pikaraoke/.venv/lib/python3.11/site-packages/selenium/webdriver/chrome/webdriver.py", line 45, in __init__
    super().__init__(
  File "/home/pi/pikaraoke/.venv/lib/python3.11/site-packages/selenium/webdriver/chromium/webdriver.py", line 49, in __init__
    self.service.path = DriverFinder.get_path(self.service, options)
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/pi/pikaraoke/.venv/lib/python3.11/site-packages/selenium/webdriver/common/driver_finder.py", line 44, in get_path
    raise NoSuchDriverException(f"Unable to locate or obtain driver for {options.capabilities['browserName']}")
selenium.common.exceptions.NoSuchDriverException: Message: Unable to locate or obtain driver for chrome; For documentation on this error, please visit: https://www.selenium.dev/documentation/webdriver/troubleshooting/errors/driver_location`

**Comment by vicwomg** [comment]
Hmm, it looks like my last update broke debian detection on the pi.
Therefore, it is skipping the setup steps that would install chromedriver.
Can you run this command and let me know the output?

cat /etc/os-release

**Reply by lsibarra** [reply]
Hi @vicwomg 
Sorry for the delay. Here is my raspberry pi 3 OS release
`PRETTY_NAME="Raspbian GNU/Linux 12 (bookworm)"
NAME="Raspbian GNU/Linux"
VERSION_ID="12"
VERSION="12 (bookworm)"
VERSION_CODENAME=bookworm
ID=raspbian
ID_LIKE=debian
HOME_URL="http://www.raspbian.org/"
SUPPORT_URL="http://www.raspbian.org/RaspbianForums"
BUG_REPORT_URL="http://www.raspbian.org/RaspbianBugs"`

**Reply by vicwomg** [reply]
Thanks for that output. I see the issue with the setup script and pushed a fix, try to `git pull` the latest version then rerun ./setup.sh, should work now.

**Reply by lsibarra** [reply]
Hi @vicwomg 
I updated and ran `git pull` the latest version, deleted .venv folder and re-ran the `./setup.sh`. It is working now. It did install the chromedriver correctly as well as recognize that the pi is running debian which also installed ffmpeg. In additon, running `./pikaraoke.sh` also automatically does the preferred ip. Everything is good now.

Thanks again for your help. If you need me to test the updated version on Windows, please let me know.

**Comment by q2000drum** [comment]
Hi,
I am also having issues with this. I already switched it to the Bookworm OS in Raspberry Pi 4. Just to be clear I installed Bookworm with Desktop and Recommended software and ./setup.sh  However after running ./pikaraoke.sh 
I am still getting the selenium chromedriver issue. See below.

pi@pikaraoke:~/pikaraoke $ ./pikaraoke.sh
Traceback (most recent call last):
  File "/home/pi/pikaraoke/app.py", line 20, in <module>
    from selenium import webdriver
  File "/home/pi/pikaraoke/.venv/lib/python3.11/site-packages/selenium/webdriver/__init__.py", line 20, in <module>
    from .chrome.webdriver import WebDriver as Chrome  # noqa
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/pi/pikaraoke/.venv/lib/python3.11/site-packages/selenium/webdriver/chrome/webdriver.py", line 18, in <module>
    from selenium.webdriver.chromium.webdriver import ChromiumDriver
  File "/home/pi/pikaraoke/.venv/lib/python3.11/site-packages/selenium/webdriver/chromium/webdriver.py", line 22, in <module>
    from selenium.webdriver.remote.webdriver import WebDriver as RemoteWebDriver
  File "/home/pi/pikaraoke/.venv/lib/python3.11/site-packages/selenium/webdriver/remote/webdriver.py", line 67, in <module>
    from .websocket_connection import WebSocketConnection
  File "/home/pi/pikaraoke/.venv/lib/python3.11/site-packages/selenium/webdriver/remote/websocket_connection.py", line 23, in <module>
    from websocket import WebSocketApp
ModuleNotFoundError: No module named 'websocket'

Cheers
